components/JoinForm/JoinForm.tsx

```diff
@@ -150,20 +153,54 @@ export default function App() {
   };
 
   async function submitJoinFormToMeshDB(joinFormSubmission: JoinFormValues) {
-    console.debug(JSON.stringify(joinFormSubmission));
-
+    // Before we try anything else, submit to S3 for safety.
+    let record: JoinRecord = Object.assign(
+      structuredClone(joinFormSubmission),
+      {
+        submission_time: new Date().toISOString(),
+        code: "",
+        replayed: 0,
+        install_number: null,
+      },
+    ) as JoinRecord;
+
+    saveJoinRecordToS3(record, joinRecordKey).then((key) => {
+      setJoinRecordKey(key as string);
+    });
+
+    // Attempt to submit the join form
     return fetch(`${await getMeshDBAPIEndpoint()}/api/v1/join/`, {
       method: "POST",
       body: JSON.stringify(joinFormSubmission),
     })
       .then(async (response) => {
```

Comment:
Any chance you could unpack this nested logic into top-level `await`s with normal `try catch` for error handling? I think the return statements and nesting are a little confusing. You might also consider putting the S3 upload in `finally` block

Reply:
I agree. I was fiddling with it the other day but couldn't get the types right. I can take another crack at it.

Reply:
I didn't end up putting the join record in the finally block. Right now my system is capable of capturing the JoinRecord before doing anything else, so even if meshdb is hard down (container is not running) we get a record. Working on writing some tests and trying to transparently handle it so we can still say "thanks, we'll get back to you in a few DAYS"